crates/rethnet_evm/src/miner.rs

```diff
@@ -0,0 +1,232 @@
+        mem_pool: Arc<RwLock<MemPool>>,
+        prevrandao_generator: RandomHashGenerator,
+        cfg: CfgEnv,
+        block_gas_limit: U256,
+        beneficiary: Address,
+    ) -> Self {
+        Self {
+            blockchain,
+            state,
+            mem_pool,
+            prevrandao_generator,
+            cfg,
+            block_gas_limit,
+            beneficiary,
+        }
+    }
+
+    /// Mines a block using as many transactions as can fit in it.
+    pub async fn mine_block(
+        &mut self,
+        timestamp: U256,
+        reward: U256,
+        base_fee: Option<U256>,
+    ) -> Result<MineBlockResult, MineBlockError<BE, SE>> {
+        let mut block_builder = {
+            let blockchain = self.blockchain.read().await;
+            let parent_block = blockchain
+                .last_block()
+                .map_err(MineBlockError::Blockchain)?;
+
+            BlockBuilder::new(
+                self.blockchain.clone(),
+                self.state.clone(),
+                self.cfg.clone(),
+                parent_block.header.clone(),
+                BlockOptions {
+                    beneficiary: Some(self.beneficiary),
+                    number: Some(parent_block.header.number),
+                    gas_limit: Some(self.block_gas_limit),
+                    timestamp: Some(timestamp),
+                    mix_hash: if self.cfg.spec_id >= SpecId::MERGE {
+                        Some(self.prevrandao_generator.next_value())
+                    } else {
+                        None
+                    },
+                    nonce: Some(if self.cfg.spec_id >= SpecId::MERGE {
+                        B64::ZERO
+                    } else {
+                        B64::from(U64::from(42))
+                    }),
+                    base_fee: if self.cfg.spec_id >= SpecId::LONDON {
+                        Some(
+                            base_fee
+                                .unwrap_or_else(|| calculate_next_base_fee(&parent_block.header)),
+                        )
+                    } else {
+                        None
+                    },
+                    ..Default::default()
+                },
+            )
+            .await?
+        };
+
+        let mut transaction_pool = self.mem_pool.write().await;
+        let mut pending_transactions: VecDeque<_> = transaction_pool
+            .pending_transactions()
+            .iter()
+            .cloned()
+            .collect();
+
+        let mut results = Vec::new();
+        let mut traces = Vec::new();
+
+        while let Some(transaction) = pending_transactions.pop_front() {
+            let mut tracer = TraceCollector::default();
+
+            let transaction_hash = *transaction.hash();
+
+            match block_builder
+                .add_transaction(transaction, Some(&mut tracer))
+                .await
+            {
+                // Question: Should we check this condition manually, to minimise overhead?
```

Comment:
We're relying on `revm` to tell us that we exceed the block gas limit. We could check this manually beforehand. What do you think is better @fvictorio & @alcuadrado ?

Reply:
Don't block on this, but bring it up with Dragan as a possible improvement - by making a strongly typed `TxEnv` that does the validation upon construction.